File: 10-wrappers/10-2-Python/10-2-8-dynamic-points-aitken/heat-right.py
"""
##############################################################################
# Multiscale Universal Interface Code Coupling Library                       #
#                                                                            #
# Copyright (C) 2023 W. Liu                                                  #
#                                                                            #
# This software is jointly licensed under the Apache License, Version 2.0    #
# and the GNU General Public License version 3, you may use it according     #
# to either.                                                                 #
#                                                                            #
# ** Apache License, version 2.0 **                                          #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License");            #
# you may not use this file except in compliance with the License.           #
# You may obtain a copy of the License at                                    #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#                                                                            #
# ** GNU General Public License, version 3 **                                #
#                                                                            #
# This program is free software: you can redistribute it and/or modify       #
# it under the terms of the GNU General Public License as published by       #
# the Free Software Foundation, either version 3 of the License, or          #
# (at your option) any later version.                                        #
#                                                                            #
# This program is distributed in the hope that it will be useful,            #
# but WITHOUT ANY WARRANTY; without even the implied warranty of             #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              #
# GNU General Public License for more details.                               #
#                                                                            #
# You should have received a copy of the GNU General Public License          #
# along with this program.  If not, see <http://www.gnu.org/licenses/>.      #
##############################################################################
#
# @file heat-right.py
# @author W. Liu
# @date 19 March 2019
# @brief Right domain on Aitken's dynamic points to demonstrate MUI coupling algorithm.
#
"""

from mpi4py import MPI
import mui4py
import numpy as np
import csv
import os

#                             Left Domain                       Right Domain             
# Coarse : +-------+-------+-------+-------o=======+=======o-------+-------+-------+-------+
#          0       1       2       3       4       5       6       7       8       9      10
# +: grid points
# o: interface points
# -: single domain zone
# =: overlapping zone

# MUI parameters
dimensionMUI = 1
data_types = {"u0": mui4py.FLOAT64,
                "u": mui4py.FLOAT64}

# MUI interface creation
config3d = mui4py.Config(dimensionMUI, mui4py.FLOAT64)
URI = "mpi://right/ifs" 
iface = mui4py.Uniface(uri=URI, config=config3d) 
iface.set_data_types(data_types)

# App common world claims 
MPI_COMM_WORLD = mui4py.mpi_split_by_app()

# Local communicator rank and size
rank = MPI_COMM_WORLD.Get_rank()
size = MPI_COMM_WORLD.Get_size()

u1 = np.zeros(110)
u2 = np.zeros(110)

# Initialise values from file
content = []
try:
    with open('Resources/right_AITKEN.csv', 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            content.append(row)
        for i in range(40, 110, 10):
            u1[i] = float(content[int(i*0.1-3)][1])
except FileNotFoundError:
    print("right_AITKEN.csv missing")
    for i in range(40, 110, 10):
        u1[i] = float(0.0)

# Create folder if don't exists
makedirMString = "results_right" + str(rank)
if not os.path.exists(makedirMString):
    os.makedirs(makedirMString)
    print("Folder " , makedirMString, " created ")
else:
    print("Folder " , makedirMString,  " already exists")
makedirMIterString = "results_iteration_right" + str(rank)
if not os.path.exists(makedirMIterString):
    os.makedirs(makedirMIterString)
    print("Folder " , makedirMIterString, " created ")
else:
    print("Folder " , makedirMIterString,  " already exists")

k, H = 0.515, 1.0

ptsVluInit = []

for i in range(40, 110, 10):
    pt = iface.Point([i])
    ptsVluInit.append((pt, u1[i]))

# Samplers
t_sampler = mui4py.TemporalSamplerExact()
s_sampler = mui4py.SamplerPseudoNearestNeighbor(30)
a_algorithm = mui4py.AlgorithmAitken(0.01, 1.0, MPI_COMM_WORLD, ptsVluInit, 0.0, config=config3d)

# Print off a hello world message
print("Hello world from Right rank ", rank, " out of ", size, " MUI processors\n")

# Output
filenameR = f"results_right{rank}/solution-right_AITKEN_0.csv"
with open(filenameR, "w") as outputFileRight:
    outputFileRight.write("\"X\",\"u\"\n")
    for i in range(40, 110, 10):
        outputFileRight.write(f"{i*H},{u1[i]},\n")
filenameIterR = f"results_iteration_right{rank}/solution-right_AITKEN_0.csv"
with open(filenameIterR, "w") as outputFileIterRight:
    outputFileIterRight.write("\"X\",\"u\"\n")
    for i in range(40, 110, 10):
        outputFileIterRight.write(f"{i*H},{u1[i]},\n")

for t in range(1, 11):
    for iter in range(1, 101):
        print(f"Right grid time {t} iteration {iter}")
        u1[40] = iface.fetch("u", 40 * H, t, iter, s_sampler, t_sampler, a_algorithm)
        
        if ((t>=4) and (t<6)):
            u1[42] = iface.fetch("u", 42 * H, t, iter, s_sampler, t_sampler, a_algorithm)

        print(f"Right under relaxation factor at t= {t} iter= {iter} is {a_algorithm.get_under_relaxation_factor(t, iter)}")
        print(f"Right residual L2 Norm at t= {t} iter= {iter} is {a_algorithm.get_residual_L2_Norm(t, iter)}")

        # calculate 'interior' points
        for i in range(50, 100, 10):
            u2[i] = u1[i] + k / (H * H) * (u1[i - 10] + u1[i + 10] - 2 * u1[i])
        # calculate 'boundary' points
        u2[100] = 0.0
        u2[40] = u1[40]

        if ((t>=4) and (t<6)):
            u2[42] = u1[42]

        # push data to the other solver
        iface.push("u0", 60 * H, u1[60])
        iface.commit(t, iter)

        # I/O
        u1, u2 = u2, u1

        # Output
        filenameR = f"results_iteration_right{rank}/solution-right_AITKEN_{((t-1)*100) + iter}.csv"
        with open(filenameR, "w") as outputFileRight:
            outputFileRight.write("\"X\",\"u\"\n")
            outputFileRight.write(f"{40*H},{u1[40]},\n")
            if ((t>=4) and (t<6)):
                outputFileRight.write(f"{42*H},{u1[42]},\n")
            for i in range(50, 110, 10):
                outputFileRight.write(f"{i*H},{u1[i]},\n")

    # Output
    filenameR = f"results_right{rank}/solution-right_AITKEN_{t}.csv"
    with open(filenameR, "w") as outputFileRight:
        outputFileRight.write("\"X\",\"u\"\n")
        outputFileRight.write(f"{40*H},{u1[40]},\n")
        if ((t>=4) and (t<6)):
            outputFileRight.write(f"{42*H},{u1[42]},\n")
        for i in range(50, 110, 10):
            outputFileRight.write(f"{i*H},{u1[i]},\n")

File: 09-algorithms/9-1-Aitken/9-1-0-basic/heat-right.cpp
/*****************************************************************************
* Multiscale Universal Interface Code Coupling Library Demo 9-1-0            *
*                                                                            *
* Copyright (C) 2023 W. Liu                                                  *
*                                                                            *
* This software is jointly licensed under the Apache License, Version 2.0    *
* and the GNU General Public License version 3, you may use it according     *
* to either.                                                                 *
*                                                                            *
* ** Apache License, version 2.0 **                                          *
*                                                                            *
* Licensed under the Apache License, Version 2.0 (the "License");            *
* you may not use this file except in compliance with the License.           *
* You may obtain a copy of the License at                                    *
*                                                                            *
* http://www.apache.org/licenses/LICENSE-2.0                                 *
*                                                                            *
* Unless required by applicable law or agreed to in writing, software        *
* distributed under the License is distributed on an "AS IS" BASIS,          *
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
* See the License for the specific language governing permissions and        *
* limitations under the License.                                             *
*                                                                            *
* ** GNU General Public License, version 3 **                                *
*                                                                            *
* This program is free software: you can redistribute it and/or modify       *
* it under the terms of the GNU General Public License as published by       *
* the Free Software Foundation, either version 3 of the License, or          *
* (at your option) any later version.                                        *
*                                                                            *
* This program is distributed in the hope that it will be useful,            *
* but WITHOUT ANY WARRANTY; without even the implied warranty of             *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
* GNU General Public License for more details.                               *
*                                                                            *
* You should have received a copy of the GNU General Public License          *
* along with this program.  If not, see <http://www.gnu.org/licenses/>.      *
******************************************************************************/

/**
 * @file heat-right.cpp
 * @author W. Liu
 * @date 16 July 2019
 * @brief Right domain on basic Aitken's to demonstrate MUI coupling algorithm.
 */

#include "mui.h"
#include <algorithm>
#include <fstream>
#include <sstream>
#include <sys/stat.h>

/*                             Left Domain                       Right Domain             
 * Coarse : +-------+-------+-------+-------o=======+=======o-------+-------+-------+-------+
 *          0       1       2       3       4       5       6       7       8       9      10
 * +: grid points
 * o: interface points
 * -: single domain zone
 * =: overlapping zone
 */

// USAGE: mpirun -np 1 ./heat-left : -np 1 ./heat-right

int main( int argc, char ** argv ) {
    using namespace mui;

    const static int N = 11;
    double u1[N], u2[N]; 

    /// Initialise values from file
    std::string inoutFilenameL = "Resources/right_AITKEN.csv";
    std::fstream inFile;
    std::vector<std::vector<std::string>> content;
    std::vector<std::string> row;
    std::string line, word;
    inFile.open(inoutFilenameL, std::ios::in);
    if(inFile.is_open())
    {
        while(getline(inFile, line)) {
            row.clear();
            std::stringstream str(line);
            while (std::getline(str, word, ',')) {
                row.push_back(word);
           }
           content.push_back(row);
        }

        for ( int i = 4; i <  11; i++ ) u1[i] = stod(content[i-3][1]);

    } else {
        std::cerr<<"right_AITKEN.csv missing" << std::endl;
        for ( int i = 4; i <  11; i++ ) u1[i] = 0.0;
    }

    inFile.close();

    uniface1d interface( "mpi://right/ifs" );

    MPI_Comm  world = mui::mpi_split_by_app();
    MPI_Comm*  Cppworld = &world;
    int rankLocal = MPI::COMM_WORLD.Get_rank();
    int sizeLocal = MPI::COMM_WORLD.Get_size();
    
    int rank, size;
    MPI_Comm_rank( world, &rank );
    MPI_Comm_size( world, &size );

    /// Create folder
    std::string makedirMString = "results_right" + std::to_string(rank);
    mkdir(makedirMString.c_str(), 0777);
    std::string fileAddress(makedirMString);
    std::string makedirMIterString = "results_iteration_right" + std::to_string(rank);
    mkdir(makedirMIterString.c_str(), 0777);
    std::string fileAddressIter(makedirMIterString);


    double        k = 0.515, H = 1;
    double *      u = u1, *v = u2;

    std::vector<std::pair<mui::point1d, double>> ptsVluInit;

    for ( int i = 4; i <  11; i++ ) {
        mui::point1d pt(i);
        ptsVluInit.push_back(std::make_pair(pt,u1[i]));
    }

    // fetch data from the other solver
    sampler_pseudo_nearest_neighbor1d<double> s1(0.1);
    temporal_sampler_exact1d  s2;
    algo_aitken1d aitken(0.01,1.0);

     // Print off a hello world message
    printf("Hello world from Right rank %d out of %d MUI processors\n",
           rank, size);
           
     // Print off a hello world message
    printf("Hello world from Right rank %d out of %d local processors\n",
           rankLocal, sizeLocal);

    /// Output
    std::ofstream outputFileRight;
    std::string filenameR = "results_right" + std::to_string(rank) + "/solution-right_AITKEN_0.csv";
    outputFileRight.open(filenameR);
    outputFileRight << "\"X\",\"u\"\n";
    for ( int i = 4; i <  11; i++ ) outputFileRight << i * H << "," << u[i] << ", \n";
    outputFileRight.close();

    std::ofstream outputFileIterRight;
    std::string filenameIterR = "results_iteration_right" + std::to_string(rank) + "/solution-right_AITKEN_0.csv";
    outputFileIterRight.open(filenameIterR);
    outputFileIterRight << "\"X\",\"u\"\n";
    for ( int i = 4; i <  11; i++ ) outputFileIterRight << i * H << "," << u[i] << ", \n";
    outputFileIterRight.close();
    for ( int t = 1; t <= 10; ++t ) {
		for ( int iter = 1; iter <= 100; ++iter ) {
			printf( "Right grid time %d iteration %d\n", t, iter );

				u[4] = interface.fetch( "u", 4 * H,  t, iter, s1, s2 , aitken);
				printf( "Right under relaxation factor at t= %d iter= %d is %f\n", t, iter, aitken.get_under_relaxation_factor(t,iter));
				printf( "Right residual L2 Norm at t= %d iter= %d is %f\n", t, iter, aitken.get_residual_L2_Norm(t,iter));
				// calculate 'interior' points
				for ( int i = 5; i <  11; i++ ) v[i] = u[i] + k / ( H * H ) * ( u[i - 1] + u[i + 1] - 2 * u[i] );
				// calculate 'boundary' points
				v[N - 1] = 0.0;
				v[4]     = u[4    ];

				// push data to the other solver
				interface.push( "u0", 6 * H, u[6] );
				interface.commit(  t, iter );
			// I/O
			std::swap( u, v );

			/// Output
			std::ofstream outputFileIterRight;
			std::string filenameIterR = "results_iteration_right" + std::to_string(rank) + "/solution-right_AITKEN_"
			  + std::to_string(((t-1)*100) + iter) + ".csv";
			outputFileIterRight.open(filenameIterR);
			outputFileIterRight << "\"X\",\"u\"\n";
			for ( int i = 4; i <  11; i++ ) outputFileIterRight << i * H << "," << u[i] << ", \n";
			outputFileIterRight.close();
		}
		/// Output
		std::ofstream outputFileRight;
		std::string filenameR = "results_right" + std::to_string(rank) + "/solution-right_AITKEN_"
		  + std::to_string(t) + ".csv";
		outputFileRight.open(filenameR);
		outputFileRight << "\"X\",\"u\"\n";
		for ( int i = 4; i <  11; i++ ) outputFileRight << i * H << "," << u[i] << ", \n";
		outputFileRight.close();
    }

    return 0;
}
